rs-matter/src/data_model/core.rs

```diff
@@ -15,142 +15,759 @@
-                    // should cause all ACLs of that fabric to be deleted and the new one to be added (Case 1).
-                    //
-                    // This is in conflict with the immediate-effect expectation of ACL: an ACL
-                    // write should instantaneously update the ACL so that immediate next WriteAttribute
-                    // *in the same WriteRequest* should see that effect (Case 2).
-                    //
-                    // As with the C++ SDK, here we do all the ACLs checks first, before any write begins.
-                    // Thus we support the Case1 by doing this. It does come at the cost of maintaining an
-                    // additional list of expanded write requests as we start processing those.
-                    let node = metadata.node();
-                    let write_attrs: heapless::Vec<_, MAX_WRITE_ATTRS_IN_ONE_TRANS> =
-                        node.write(req, &accessor).collect();
-
-                    for item in write_attrs {
-                        AttrDataEncoder::handle_write(&item, &self.0, &mut driver.writer()?)
-                            .await?;
-                    }
+            });
+
+            let primed = self
+                .report_data(id, node_id, &rx, &mut tx, exchange)
+                .await?;
+
+            if primed {
+                exchange
+                    .send_with(|_, wb| {
+                        SubscribeResp::write(wb, id, max_int_secs)?;
+                        Ok(Some(OpCode::SubscribeResponse.into()))
+                    })
+                    .await?;
 
-                    driver.complete(req).await?;
+                info!("Subscription {node_id:x}::{id} created");
+
+                if self.subscriptions.update(id, node_id) {
+                    let _ = self
+                        .subscriptions_buffers
+                        .borrow_mut()
+                        .push(SubscriptionBuffer {
+                            node_id,
+                            id,
+                            buffer: rx,
+                        });
+
+                    subscribed.set(true);
                 }
-                Interaction::Invoke {
-                    req,
-                    ref mut driver,
-                } => {
-                    let accessor = driver.accessor()?;
+            }
+        } else {
+            Self::send_status(exchange, IMStatusCode::ResourceExhausted).await?;
+        }
 
-                    for item in metadata.node().invoke(req, &accessor) {
-                        let (mut tw, exchange) = driver.writer_exchange()?;
+        Ok(())
+    }
 
-                        CmdDataEncoder::handle(&item, &self.0, &mut tw, exchange).await?;
-                    }
+    pub async fn process_subscriptions(&self, matter: &Matter<'_>) -> Result<(), Error> {
+        loop {
+            let mut timeout = pin!(Timer::after(embassy_time::Duration::from_secs(4)));
```

Comment:
where does the 4 seconds number come from? Could we have this as a const so we document it as well? A magic number somewhere in the middle of the file seems a bit less obvious.

Reply:
I expect the subscription notification code to change with new PRs. This would also most likely affect these 4 seconds which are rather arbitrary and probably need a more involved computation, based on min/max report time of all active subscriptions.

For now, I've left a TODO.